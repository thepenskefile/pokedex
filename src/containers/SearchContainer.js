// @flow
import React, { Component } from 'react';
import SearchResultsComponent from '../components/SearchResultsComponent';
import axios from 'axios';
import Loads from 'react-loads';
import { Box, Spinner } from 'fannypack';
import { navigate } from '@reach/router';
import { CONTENT_CATEGORY_TYPES } from '../_types/content_category_types';

type Props = {
  category: string
};

export default class SearchContainer extends Component<Props> {
  loadOptions = async ({ page, searchText }: { page: number, searchText: string }) => {
    const limit = 30;
    const response = await axios.get(
      `https://pokeapi.co/api/v2/${this.props.category}/?limit=${limit}&offset=${limit * (page - 1)}`
    );
    let options = null;
    if (searchText) {
      options = response.data.results.map((item: any) => {
        if (item.name.includes(searchText)) return { key: item.name, label: item.name, value: item };
      });
      options = options.filter(function(element) {
        return element !== undefined;
      });

      while (options.length === 0) {
        const searchPagesLimit = 300;

        var searchPages = page;
        var data = await axios.get(
          `https://pokeapi.co/api/v2/${this.props.category}/?limit=${searchPagesLimit}&offset=${searchPagesLimit *
            (searchPages++ - 1)}`
        );

        options = data.data.results.map((item: any) => {
          if (item.name.includes(searchText)) return { key: item.name, label: item.name, value: item };
        });
        options = options.filter(function(element) {
          return element !== undefined;
        });
        if (data.next === undefined) {
          break;
        }
      }
      // $FlowFixMe
      options = options.filter(function(element) {
        return element !== undefined;
      });
    } else {
      options = response.data.results.map((item: any) => ({ key: item.name, label: item.name, value: item }));
    }
    return { options };
  };

  handleClickMenuItem = (id: string) => {
    navigate(`/${CONTENT_CATEGORY_TYPES[this.props.category]}/${id}`);
  };
  render = () => {
    return (
      <Loads contextKey={this.props.category}>
        {({ isLoading }) => (
          <Box marginLeft="7px" marginRight="7px">
            {isLoading && (
              <Box textAlign="center">
                <Spinner margin="5px" marginTop="20px" textAlign="center" size="large" color="text" />
              </Box>
            )}
            <SearchResultsComponent loadOptions={this.loadOptions} onClickMenuItem={this.handleClickMenuItem} />
          </Box>
        )}
      </Loads>
    );
  };
}
