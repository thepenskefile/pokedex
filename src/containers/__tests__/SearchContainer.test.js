import React from 'react';
import SearchContainer from '../SearchContainer';
import renderer from 'react-test-renderer';

const defaultProps = {};

it('renders correctly', () => {
  const component = renderer.create(<SearchContainer {...defaultProps} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
