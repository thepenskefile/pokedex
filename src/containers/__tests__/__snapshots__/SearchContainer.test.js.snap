// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Box-sc-6sssmg-0 inpIDs"
  style={
    Object {
      "marginLeft": "7px",
      "marginRight": "7px",
    }
  }
>
  <div
    className="Field-jd7v3e-0 styled-ixmo9v-5 iZvzCO Box-sc-6sssmg-0 inpIDs"
  >
    <div
      className="sc-bdVaJa kulbDP styled__InputWrapper-kl4ffb-2 Rsbvv Box-sc-6sssmg-0 inpIDs"
    >
      <input
        aria-invalid={false}
        aria-required={false}
        autoFocus={false}
        className="styled-kl4ffb-3 caPFUM Input-tpacev-0 igsxQX Box-sc-6sssmg-0 inpIDs"
        disabled={false}
        multiple={false}
        placeholder="search"
        type="text"
      />
    </div>
    <div />
  </div>
  <div
    className="styled-sc-144wh3b-0 iUlbjp Box-sc-6sssmg-0 inpIDs"
    style={
      Object {
        "marginTop": "20px",
        "textAlign": "left",
      }
    }
  >
    <div
      className="Box-sc-6sssmg-0 inpIDs"
      height="70vh"
      onScroll={[Function]}
      style={
        Object {
          "height": "70vh",
          "overflowX": "hidden",
          "overflowY": "scroll",
        }
      }
    />
  </div>
</div>
`;
